File: jstest/recovery_test.js
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import dataShare from '@ohos.data.dataShare';
import {afterAll, afterEach, beforeAll, beforeEach, describe, expect, it} from 'deccjsunit/index'

const URI_CONTACTS = "datashare:///com.ohos.contactsdataability";
const rawContactUri = "datashare:///com.ohos.contactsdataability/contacts/raw_contact";
const backup = "datashare:///com.ohos.contactsdataability/contacts/backup";
const recover = "datashare:///com.ohos.contactsdataability/contacts/recover";
const deletedUri = "datashare:///com.ohos.contactsdataability/contacts/deleted_raw_contact";

describe('RecoveryTest', function() {
    function sleep(numberMillis)
    {
        var now = new Date();
        var exitTime = now.getTime() + numberMillis;
        while (true) {
            now = new Date();
            if (now.getTime() > exitTime)
                return;
        }
    }

    async function ContactsQuery(tag, uri, size)
    {
        let dataShareHelper = dataShare.createDataShareHelper(URI_CONTACTS);
        console.info(tag + ': ContactsQuery start ! dataShareHelper = ' + dataShareHelper);
        var resultColumns = [];
        let condition = new dataShare.DataSharePredicates();
        condition.greaterThan("id", "0");
        try {
            var resultSet = await dataShareHelper.query(uri, resultColumns, condition);
            console.info(tag + ' ContactsQuery resultSet.rowCount is = ' + resultSet.rowCount);
            console.info(tag + ' ContactsQuery size is = ' + size);
            expect(resultSet.rowCount == size).assertEqual(true);
            if (resultSet.rowCount > 0) {
                if (resultSet.goToFirstRow()) {
                    do {
                        var id = resultSet.getString(resultSet.getColumnIndex("id"));
                        console.info(tag + ':logMessage contactsQuery id = ' + id);
                        var name = resultSet.getString(resultSet.getColumnIndex("display_name"));
                        console.info(tag + ':logMessage contactsQuery name = ' + name);
                    } while (resultSet.goToNextRow())
                }
            }
            resultSet.close();
        } catch (error) {
            console.info(tag + "logMessage ContactsQuery: error = " + error);
        }
    }

    async function rawContactInsert(name)
    {
        console.info("------logMessage rawContactInsert is starting!-----");
        let dataShareHelper = dataShare.createDataShareHelper(URI_CONTACTS);
        console.info('logMessage get dataShareHelper success! dataShareHelper = ' + dataShareHelper);
        var insertValue = {
            "display_name" : name,
        }
        try {
            var rawContactId = await dataShareHelper.insert(rawContactUri, insertValue);
            console.info('logMessage rawContactInsert: rawContactId = ' + rawContactId);
            expect(rawContactId > 0).assertTrue();
        } catch (error) {
            console.info('logMessage rawContactInsert: insert error = ' + error);
        }
        return rawContactId;
    }

    async function deleteRawContact(tag)
    {
        try {
            let dataShareHelper = dataShare.createDataShareHelper(URI_CONTACTS);
            let condition = new dataShare.DataSharePredicates();
            condition.greaterThan("id", "0");
            condition.and();
            condition.equalTo("is_deleted", "0");
            await dataShareHelper.delete(rawContactUri, condition);
            sleep(5000);
            var conditionAll = new dataShare.DataSharePredicates();
            conditionAll.greaterThan("id", "0");
            await dataShareHelper.delete(deletedUri, conditionAll);
        } catch (error) {
            console.info(tag + ': deleteRawContact error = ' + error);
        }
    }

    /**
     * @tc.number  recovery_test_100
     * @tc.name    backup database
     * @tc.desc    Function test
     */
    it("recovery_test_100", 0, async function(done) {
        await deleteRawContact("recovery_test_100");
        let dataShareHelper = dataShare.createDataShareHelper(URI_CONTACTS);
        var updateValues = {};
        let condition = new dataShare.DataSharePredicates();
        try {
            var updateCode = await dataShareHelper.update(backup, updateValues, condition);
            sleep(2000);
            console.info('logMessage recovery_test_100: updateCode = ' + updateCode);
            expect(updateCode == 0).assertTrue();
            sleep(2000);
            await ContactsQuery("recovery_test_100", rawContactUri, 0);
            done();
        } catch (error) {
            console.info('logMessage recovery_test_100: raw_contact update error = ' + error);
            done();
        }
    });

    /**
     * @tc.number  recovery_test_200
     * @tc.n    ame    recovery database
     * @tc.desc    Function test
     */
    it("recovery_test_200", 0, async function(done) {
        await deleteRawContact("recovery_test_200");
        let dataShareHelper = dataShare.createDataShareHelper(URI_CONTACTS);
        var updateValues = {};
        let condition = new dataShare.DataSharePredicates();
        rawContactInsert("liming");
        rawContactInsert("xiaolilili");
        sleep(2000);
        var updateCode = await dataShareHelper.update(backup, updateValues, condition);
        sleep(2000);
        console.info(' recovery_test_200: backup = ' + updateCode);
        expect(updateCode == 0).assertTrue();
        sleep(1000);
        try {
            var updateCode = await dataShareHelper.update(recover, updateValues, condition);
            sleep(3000);
            console.info('logMessage recovery_test_200: recover = ' + updateCode);
            expect(updateCode == 0).assertTrue();
            await ContactsQuery("recovery_test_200", rawContactUri, 2);
            await deleteRawContact("recovery_test_200");
            done();
        } catch (error) {
            console.info('logMessage recovery_test_200: raw_contact update error = ' + error);
            done();
        }
    });
});

File: test/unittest/src/recovery_test.cpp
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "recovery_test.h"

#include "database_disaster_recovery.h"
#include "test_common.h"

namespace Contacts {
namespace Test {
void RecoveryTest::ClearData()
{
    OHOS::AbilityRuntime::ContactsDataAbility contactsDataAbility;
    OHOS::DataShare::DataSharePredicates predicates;
    OHOS::Uri uriRawContact(ContactsUri::RAW_CONTACT);
    predicates.NotEqualTo("id", "0");
    predicates.And();
    predicates.EqualTo("is_deleted", "0");
    int code1 = contactsDataAbility.Delete(uriRawContact, predicates);
    HILOG_INFO("RecoveryTest ClearData code1 is %{public}d", code1);
    int time = 3000;
    std::chrono::milliseconds dura(time);
    std::this_thread::sleep_for(dura);
    OHOS::DataShare::DataSharePredicates predicates2;
    OHOS::Uri uriRawContactComplete(ContactsUri::DELETED_RAW_CONTACT);
    predicates2.NotEqualTo("id", "0");
    int code = contactsDataAbility.Delete(uriRawContactComplete, predicates2);
    HILOG_INFO("RecoveryTest ClearData code is %{public}d", code);
}

int64_t RecoveryTest::RawContactInsert(std::string displayName)
{
    OHOS::Uri uriRawContact(ContactsUri::RAW_CONTACT);
    OHOS::DataShare::DataShareValuesBucket rawContactValues;
    rawContactValues.Put("display_name", displayName);
    int64_t code = contactsDataAbility.Insert(uriRawContact, rawContactValues);
    rawContactValues.Clear();
    return code;
}

/*
 * @tc.number  recovery_test_100
 * @tc.name    Backup database
 * @tc.desc    Exception use case
 * @tc.level   Level1
 * @tc.size    MediumTest
 * @tc.type    Function
 */
HWTEST_F(RecoveryTest, recovery_test_100, testing::ext::TestSize.Level1)
{
    HILOG_INFO("--- recovery_test_100 is starting! ---");
    std::shared_ptr<OHOS::Contacts::DataBaseDisasterRecovery> instance =
        OHOS::Contacts::DataBaseDisasterRecovery::GetInstance();
    std::shared_ptr<OHOS::Contacts::ContactsDataBase> contactDatabase = OHOS::Contacts::ContactsDataBase::GetInstance();
    int result = instance->SQLiteCheckDb();
    HILOG_INFO("--- recovery_test_100 SQLiteCheckDb ret ! --- %{public}d", result);
    if (result == 0) {
        instance->BackDatabase();
    }
    std::string backupPath = instance->GetBackUpDatabase(contactDatabase->contactStore_);
    HILOG_INFO("--- recovery_test_100 is end! --- %{private}s ", backupPath.c_str());
    ClearData();
}

/*
 * @tc.number  recovery_test_200
 * @tc.name    Backup database after insert two person, then recover
 * @tc.desc    Exception use case
 * @tc.level   Level1
 * @tc.size    MediumTest
 * @tc.type    Function
 */
HWTEST_F(RecoveryTest, recovery_test_200, testing::ext::TestSize.Level1)
{
    HILOG_INFO("--- recovery_test_200 is starting! ---");
    RawContactInsert("liming");
    RawContactInsert("xiaolilili");
    // backup
    OHOS::Uri uriRawContactBackUp(ContactsUri::BACKUP);
    OHOS::DataShare::DataShareValuesBucket value;
    OHOS::DataShare::DataSharePredicates predicates;
    int retCode = contactsDataAbility.Update(uriRawContactBackUp, predicates, value);
    EXPECT_EQ(0, retCode);

    RawContactInsert("xiaobaibaibai");
    RawContactInsert("xiaolitiantian");
    OHOS::Uri uriRawContact(ContactsUri::RAW_CONTACT);
    std::vector<std::string> columns;
    predicates.EqualTo("is_deleted", "0");
    std::shared_ptr<OHOS::DataShare::DataShareResultSet> resultSet =
        contactsDataAbility.Query(uriRawContact, predicates, columns);
    int rowCount = 0;
    resultSet->GetRowCount(rowCount);
    resultSet->Close();
    EXPECT_EQ(4, rowCount);

    OHOS::Uri uriRawContactRecover(ContactsUri::RECOVER);
    int retCodeRecover = contactsDataAbility.Update(uriRawContactRecover, predicates, value);
    EXPECT_EQ(0, retCodeRecover);

    std::shared_ptr<OHOS::DataShare::DataShareResultSet> resultSetRecover =
        contactsDataAbility.Query(uriRawContact, predicates, columns);
    int rowCountRecover = 0;
    resultSetRecover->GetRowCount(rowCountRecover);
    resultSetRecover->Close();
    EXPECT_EQ(2, rowCountRecover);
    ClearData();
}
} // namespace Test
} // namespace Contacts
